### PR TITLE
refactor: shared dependencies unification

### PR DESCRIPTION
**Pull request by mgorsk1**
### Summary of Changes

Following merging https://github.com/amundsen-io/rfcs/pull/38 this PR unifies dependencies across amundsen services.

#### Assumptions

- global unification (& upgrade to more recent vrersions) of libraries has been conducted (every repo now uses same version for pytest, flake8 and so on). if there were mulitple versions of library X across different projects (for example - we had 3 different versions of mypy - `0.660`, `0.782` and `0.812` across frontend,metadata and search) - the most recent one was selected for unification (in case of mypy - `0.812`)
- any linting issues that has arisen because of the previous point are solved by excluding certain types of errors (fixing linting in whole repo would make this PR a major pain to review and resolve)
- databuilder is excluded from requirements-common.txt unification 

### Tests

- all existing linter rules passing
- all existing unit tests passing

#### TODO

- verify correctness of building docker images

### Documentation

- updated appropriate README.md and installation.md files

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.
- [x] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

File: frontend/Makefile
Comment (bbsbb) [inline frontend/Makefile:28]: Shouldn't this file refer to Dockerfile.frontend.public ?

Comment (mgorsk1) [reply frontend/Makefile]: it should, nice catch!

File: Dockerfile.frontend.local
Comment (sewardgw) [inline Dockerfile.frontend.local:21]: Are there dev requirements needed for the container?

File: search/Makefile
Comment (sewardgw) [inline search/Makefile:55]: for each of the `install_deps` in the make files, is the `-e` strictly required when doing pip install? 

I'm trying to think if there is a developer experience where the dev may do `pip install .[all]` and if it could fail.

File: metadata/requirements-common.txt
Comment (sewardgw) [inline metadata/requirements-common.txt:1]: nice! Didn't know you could just reference files like this

**Review by sewardgw — APPROVED** [review]
Thanks for the work here! clearly a big effort for a much needed change. A few comments / questions inline

File: search/Makefile
Comment (mgorsk1) [reply search/Makefile:55]: it will fail if you skip `-e`. Without `-e` it makes a copy of a source dir to /tmp and conducts installation there. When this copy operation is conducted symlinks are not resolved so requirements-common.txt (or requirements-dev.txt) are not copied too. Adding -e flag changes the process in such way that files are not copied to /tmp beforehand but installation is conducted in source dir.

File: metadata/requirements-common.txt
Comment (mgorsk1) [reply metadata/requirements-common.txt:1]: we use it in several places in `docs` subdirectory :)

File: Dockerfile.frontend.local
Comment (mgorsk1) [reply Dockerfile.frontend.local:21]: without this file `pip install -e .[all]` fails cause `requirements-dev.txt` is missing (as `setup.py` fails on missing file when reading in to `requirements_dev` variable). 
I can either:
- remove this file from `Dockerfile` and introduce handling of missing files in `setup.py`  (load empty array to `requirements_dev` variable if the file is missing)
- keep this file in `Dockerfile` (they are not installed actually, just the file needs to be there `setup.py` to execute correctly) and not amend `setup.py` 

my preference is the latter one but if you feel differently I can make the change.

Comment (sewardgw) [reply Dockerfile.frontend.local:21]: Thanks for the context. I agree with this approach. This PR was also about unifying the requirements, which appears to have been done. Should this be an issue in the future it can be addressed then.

File: search/Makefile
Comment (sewardgw) [reply search/Makefile:55]: Got it. Trying to think through a scenario where someone makes changes / installs from source but forgets to add the `-e` flag. Is it worth updating the readme / docs somewhere to mention this?

**Review by sewardgw — APPROVED** [review]

**Comment by sewardgw** [comment]
One last thought - worth squashing some of the commits?

**Comment by mgorsk1** [comment]
> One last thought - worth squashing some of the commits?

I always `Squash and Merge` so main stays decluttered but good point, it's a messy one!

File: Dockerfile.frontend.local
Comment (mgorsk1) [reply Dockerfile.frontend.local:21]: I've updated README.md for search and metadata + `docs/installation.md` for frontend. Should I add it also somewhere else?

**Comment by mgorsk1** [comment]
I think this requires approval of codeowner of all repos to merge, are you up for review @feng-tao ?

File: databuilder/setup.py
Comment (dorianj) [inline databuilder/setup.py:10]: I thought for databuilder, we didn't want to specify exact deps (since it may be installed alongside other stuff) -- I think we should either update the requirements.txt to be less-specific, or remove this. WDYT?

**Review by dorianj — APPROVED** [review]

File: requirements-dev.txt
Comment (feng-tao) [inline requirements-dev.txt:10]: I think it could be problematic for databuilder given it is used for library not a service like other modules, we should at least make it as range instead of pin down to certain versions.

File: databuilder/setup.cfg
Comment (feng-tao) [inline databuilder/setup.cfg:16]: could you give some context what we ignore here?

File: databuilder/setup.py
Comment (feng-tao) [reply databuilder/setup.py:10]: +1 on dorian's comment

File: databuilder/setup.py
Comment (feng-tao) [inline databuilder/setup.py:81]: also why not put common into databuilder given eventually we should align model definition between common and databuilder?

**Review by feng-tao — CHANGES_REQUESTED** [review]

File: databuilder/setup.py
Comment (mgorsk1) [reply databuilder/setup.py:81]: requirements_common contain mostly shared dependencies for Flask services (+ obviously amundsen-common). during rfc we decided to exclude databuilder from sharing those (which would make sense if we decide not to pin any deps in databuilder, while in requirements_common flask and other deps are need to be pinned).

File: requirements-dev.txt
Comment (mgorsk1) [reply requirements-dev.txt:10]: ok, so `>=` for all requirements dev ?

File: databuilder/setup.py
Comment (mgorsk1) [reply databuilder/setup.py:10]: it doesn't really change the approach we had before but if you think that with this PR we want to stop pinning deps in databuilder (at least those generic ones) then I will remove version pinning.

File: databuilder/setup.cfg
Comment (mgorsk1) [reply databuilder/setup.cfg:16]: I think that with bumping flake8 it just became more strict with checks. `make lint` on `main` branch passes while here (after bumping from 3.5.0 to 3.9.2) it yields

```
(.venv) root@mariusz-XPS-13-7390-2-in-1:/amundsen/databuilder# make lint
flake8 .
./databuilder/task/task.py:72: [F541] f-string is missing placeholders
./databuilder/publisher/elasticsearch_publisher.py:69: [W504] line break after binary operator
./databuilder/publisher/elasticsearch_publisher.py:80: [E741] ambiguous variable name 'l'
./databuilder/publisher/neo4j_csv_publisher.py:460: [F541] f-string is missing placeholders
./databuilder/publisher/atlas_csv_publisher.py:89: [F541] f-string is missing placeholders
./databuilder/publisher/atlas_csv_publisher.py:104: [F541] f-string is missing placeholders
./databuilder/publisher/atlas_csv_publisher.py:217: [F541] f-string is missing placeholders
./databuilder/models/atlas_serializable.py:52: [F541] f-string is missing placeholders
./databuilder/models/schema/schema.py:29: [E252] missing whitespace around parameter equals
./databuilder/models/schema/schema.py:29: [E252] missing whitespace around parameter equals
./databuilder/models/schema/schema.py:30: [E252] missing whitespace around parameter equals
./databuilder/models/schema/schema.py:30: [E252] missing whitespace around parameter equals
./databuilder/extractor/dbt_extractor.py:124: [E117] over-indented
./databuilder/extractor/hive_table_last_updated_extractor.py:228: [F541] f-string is missing placeholders
./databuilder/extractor/delta_lake_metadata_extractor.py:42: [W504] line break after binary operator
./databuilder/extractor/delta_lake_metadata_extractor.py:43: [W504] line break after binary operator
./databuilder/extractor/delta_lake_metadata_extractor.py:44: [W504] line break after binary operator
./databuilder/extractor/delta_lake_metadata_extractor.py:45: [W504] line break after binary operator
./databuilder/extractor/delta_lake_metadata_extractor.py:46: [W504] line break after binary operator
./databuilder/extractor/atlas_search_data_extractor.py:237: [F841] local variable 'e' is assigned to but never used
./databuilder/extractor/atlas_search_data_extractor.py:354: [F541] f-string is missing placeholders
./databuilder/extractor/dashboard/apache_superset/apache_superset_chart_extractor.py:57: [F541] f-string is missing placeholders
./databuilder/extractor/dashboard/apache_superset/apache_superset_metadata_extractor.py:59: [F541] f-string is missing placeholders
./databuilder/extractor/dashboard/tableau/tableau_dashboard_last_modified_extractor.py:43: [W504] line break after binary operator
./databuilder/extractor/dashboard/tableau/tableau_dashboard_extractor.py:49: [W504] line break after binary operator
./example/scripts/sample_dbt_loader.py:68: [F541] f-string is missing placeholders
./example/scripts/sample_deltalake_metadata.py:52: [F541] f-string is missing placeholders
./example/scripts/sample_pandas_profiling_column_stats_loader.py:16: [F541] f-string is missing placeholders
./example/scripts/sample_feast_loader.py:54: [F541] f-string is missing placeholders
./tests/unit/publisher/test_mysql_csv_publisher.py:22: [F541] f-string is missing placeholders
./tests/unit/publisher/test_neo4j_csv_publisher.py:23: [F541] f-string is missing placeholders
./tests/unit/extractor/test_csv_extractor.py:24: [F541] f-string is missing placeholders
./tests/unit/extractor/test_kafka_source_extractor.py:18: [F541] f-string is missing placeholders

```

File: requirements-dev.txt
Comment (feng-tao) [reply requirements-dev.txt:10]: let's make it range .

File: databuilder/setup.cfg
Comment (feng-tao) [reply databuilder/setup.cfg:16]: could you file a ticket and see if we could get some community member to help fixing these?

Comment (mgorsk1) [reply databuilder/setup.cfg:16]: sure, will do after merging. sounds like a nice idea for `good-first-issue`

**Review by feng-tao — APPROVED** [review]

File: databuilder/setup.py
Comment (mgorsk1) [reply databuilder/setup.py:10]: I have started to work on requirements.txt in databuilder per your suggestion. I didn't find a lot of deps there that are directly mentioned in any of the `extras_require` section. There were few I've moved (to atlas and rds) and my approach for the remaining ones would be not to remove pinning deps completely but instead replace `==` with `>=` . This way we at least we don't fall into regression issues if some library requires `lower` version than we've previously supported. Let me know your thoughts on that. @feng-tao @dorianj

**Pull request merged**
